# Merge expression values generated by featurecounts  
You need to run featurecounts for each sample  
```{bash, eval = F}
cd $SNIC_TMP
cp /proj/uppstore2017083/private/SMS_5123_20_AM_fungi_annotation/intermediate/XenofilteR_GSNAP_assembly1/Sample_TB-2463-31A/Filtered_bams_AM_Fungi/Sample_TB-2463-31A_Filtered.bam $SNIC_TMP
featureCounts_path="/proj/uppstore2017083/private/SMS_5123_20_AM_fungi_annotation/intermediate/featureCounts/GSNAP_XenofilteR_assembly1//Sample_TB-2463-31A"
output="count-s-2"
annotation="/proj/uppstore2017083/private/SMS_5123_20_AM_fungi_annotation/data/meta_data/annotation_1/annotation_funannotate.gtf"
mkdir -p $featureCounts_path
~/glob/Software/subread-2.0.0-source/bin/featureCounts -s 2 -t exon -g gene_id -F GTF -C -T 20 -p -B -Q 20 -o $output -a $annotation Sample_TB-2463-31A_Filtered.bam 
mv $output* $featureCounts_path
```

Then we merge the counts  
```{r, eval = F ,echo = T}
path <- '~/SMS_5123_20_AM_fungi_annotation/'
samples_info <- read.table(paste0(path, 'code/Samples_reads.txt'))
colnames(samples_info) <- c('Sample', 'R1', 'R2')
samples_info <- as.data.frame(samples_info[,1])
colnames(samples_info) <- c('Sample')
cntr <- 1

for(s in samples_info$Sample){
  tmp_featurecounts <- read.table(paste0(path, 'intermediate/featureCounts/GSNAP_Merged_XenofilteR_assembly1/',s,'/count-s-2'), header = T, skip = 1)
  if(cntr == 1){
    data_raw <- data.frame(tmp_featurecounts[,1], matrix(0,nrow=nrow(tmp_featurecounts), ncol=nrow(samples_info), byrow = T))
    colnames(data_raw) <- c('gene_id',as.character(samples_info$Sample))
    data_raw[,s] <- tmp_featurecounts[,7]
  }else{
    data_raw[, as.character(s)] <- tmp_featurecounts[,7]
  }
  cat('\r', paste0(cntr,": ", as.character(s)))
  cntr <- cntr + 1
}
write.table(data_raw, paste0(path,'intermediate/featureCounts/GSNAP_Merged_XenofilteR_assembly1/merged_gene_counts.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
write.table(data_raw, paste0(path,'results/03-Expression-Transcriptome-Assembly/merged_gene_counts.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
```

By following code you can generate fraction of reads assigend to features  
```{r, eval = F}
path <- '~/SMS_5123_20_AM_fungi_annotation/'
samples_info <- read.table(paste0(path, 'code/Samples_reads.txt'))
colnames(samples_info) <- c('Sample', 'R1', 'R2')
samples_info <- as.data.frame(samples_info[,1])
colnames(samples_info) <- c('Sample')
cntr <- 1

my_stats <- c('Status')
my_stats_df <- data.frame(matrix(0,ncol = length(my_stats),nrow=1))     
colnames(my_stats_df) <- my_stats
my_stats_df <- my_stats_df[FALSE,]


for(s in samples_info$Sample){
  tmp_featurecounts <- read.table(paste0(path, 'intermediate/featureCounts/GSNAP_Merged_XenofilteR_assembly1/',s,'/count-s-2.summary'), header = T)
  if(cntr == 1){
    my_stats_df <- cbind.data.frame(Status = tmp_featurecounts$Status, tmp_featurecounts[,2]*100/colSums(as.matrix(tmp_featurecounts[,2])))
    colnames(my_stats_df)[ncol(my_stats_df)] <- s
  }else{
    my_stats_df <- cbind.data.frame(my_stats_df, tmp_featurecounts[,2]*100/colSums(as.matrix(tmp_featurecounts[,2])))
    colnames(my_stats_df)[ncol(my_stats_df)] <- s
  }
  cat('\r', paste0(cntr,": ", as.character(s)))
  cntr <- cntr + 1
}
write.table(my_stats_df, paste0(path,'intermediate/featureCounts/GSNAP_Merged_XenofilteR_assembly1/merged_summary.txt'), col.names = T, row.names = F, quote = F, sep = '\t')
write.table(my_stats_df, paste0(path,'results/03-Expression-Transcriptome-Assembly/merged_summary.txt'), col.names = T, row.names = F, quote = F, sep = '\t')

# rownames(my_stats_df) <- my_stats_df$Status
# colnames(my_stats_df) <- gsub('Sample_TB-2463-', '', colnames(my_stats_df))

#Removing rows with zero value/%
my_stats_df <- (my_stats_df[(rowSums(my_stats_df[,-1])>0),])

png(paste0(path,'results/03-Expression-Transcriptome-Assembly/merged_summary.png'), width = 2000, height = 2000, res = 200)
heatmap.2(as.matrix(my_stats_df[,-1]), dendrogram='none', Rowv=F, Colv=F, scale="none", density.info="none", trace="none", key=TRUE, keysize=1.2, cexCol=1, margins=c(5,15), main="Features summary")
dev.off()
```
